[RFC] Bazel (#92758)
elastic/kibana: rfcs/text/0015_bazel.md

One of the primary advantages Bazel provides is that the builds are hermetic, meaning they are dependent only on a known set of inputs to ensure the builds are reproducible and cacheable. To create these assurances, builds utilize a sandbox environment with only the defined dependencies available. This not only allows for aggressive local caching but the use of remote caching as well. If Bazel determines that a package or plugin needs to be re-built and it's not in the local cache, it will check the remote cache and, if found, will persist locally for subsequent builds. Once the project has completely migrated to Bazel, a developer will only build code they have directly modified or is dependent on those changes. In building packages and plugins, the expected cost for most developers will be downloading the builds from the remote cache for anything changed since their last build.

> Building TypeScript reference definitions and using `@babel/register` will be negated by using the TypeScript compiler directly instead of using Babel. Currently, we use Babel for code generation and tsc for type check and type declaration output. Additionally, the TypeScript implementation in [rules_nodejs](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) for Bazel handles incremental builds, resulting in faster re-builds.

Reviewer: ```suggestion
Building TypeScript reference definitions and using `@babel/register` will be negated by using the TypeScript compiler directly instead of using Babel. Currently, we use Babel for code generation and `tsc` for type check and type declaration output. Additionally, the TypeScript implementation in [rules_nodejs](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) for Bazel handles incremental builds, resulting in faster re-builds.
```

> In addition to the benefits of building code, there are also benefits regarding running unit tests. Developers currently need to understand what unit tests to run to validate changes or rely on waiting for CI, which has a long feedback loop. Since Bazel knows the dependency tree, it will only run unit tests for a package or plugin modified or dependent on those modifications. This optimization helps developers and will significantly reduce the amount of work CI needs to do.

Author: 423b191488f93c12a78e40e2efa4ba64b4b57990

> In addition to the benefits of building code, there are also benefits regarding running unit tests. Developers currently need to understand what unit tests to run to validate changes or rely on waiting for CI, which has a long feedback loop. Since Bazel knows the dependency tree, it will only run unit tests for a package or plugin modified or dependent on those modifications. This optimization helps developers and will significantly reduce the amount of work CI needs to do.

Author: c21f8c238591ecd261bca0acbc025513a48fe7d1

## Packages Build Outline

Within Bazel, the packages will have new overall rules:
 
* It cannot contain build scripts. Every package build will be written using a Bazel `BUILD.bazel` file
* It cannot have side effects. Every package build should be cacheable and reproducible and can not produce any side effects
* Each package should define three major public target rules in `BUILD.bazel` files: `build`, `jest`, and a js_library target with the same name of the folder where the package is living. 
* In order to output its targets in the most Bazel friendly way, each package will output its target according to the following folder structure: for node targets, it will be `target_server`, for web target it will be `target_web` and for types, it will be `target_types`.

## package.json’s Outline

As a prerequisite for Bazel and for additional benefits outlined in pull-request [#76412](https://github.com/elastic/kibana/issues/76412), the Kibana repository went from using Yarn Workspaces to a single `package.json` defining all dependencies.

> One of the benefits Bazel has over Gradle is the support for Node modules. Bazel will manage the dependencies using either the NPM or Yarn Package Manager. When doing this, a `BUILD.bazel` file will be generated for each module allowing for fine-grained control.

Reviewer 2: I'm not terribly familiar with the specifics, so this section is a bit unclear to me. What sort of fine-grained control will we have with Bazel that we don't have today?

Bonus question: will Bazel allow us to move client-side production dependencies out of `devDependencies`, or will they continue to exist as a `devDependency` for the foreseeable future?

Reviewer 3: Regarding the first question, to build a certain package Bazel force us to declare every dependency. That way the build will only run if some of your input dependencies have changed since the last build. When you transport this to node_modules there are 2 ways to look at it: or you consider `node_modules` as a closed box and if you depend on anything from there you always have to rebuild once a single file change, or you clearly know what is the composition of each module inside node_modules. To get the benefits of the latest, during yarn install Bazel will generate files that will allow the tool to know what is inside each node_module.

For the second question, @tylersmalley I think we can already do it as we have a build task that will remove non used dependencies on server code from the distributable package.json right? Or am I missing something?

Author: Yeah, I _think_ we can do that today.

Reviewer 2: Thanks for explaining, that makes more sense to me now

    ".npmignore",
    "index.js",
    "index.d.ts",
    "package.json",
    "readme",
    "tsconfig.json",
]

filegroup(
    name = "src",
    srcs = glob(SRCS),
)

js_library(
    name = "elastic-datemath",
    srcs = [ ":src" ],
    deps = [ "@npm//moment" ],
    package_name = "@elastic/datemath",
    visibility = ["//visibility:public"],
)

alias(
    name = "build",
    actual = "elastic-datemath",
    visibility = ["//visibility:public"],
)
```

If the package has unit tests, they will need to be migrated which will be invoked with `bazel test` as described in the Unit Testing section.

## Phase II - Docs, Developer Experience

Packages were a likely choice for phase 1 for a few reasons; they aren’t often updated and the developer experience is quite lacking making it easy to maintain parity with. In phase 2, we will bring the developer experience of packages to that which developers are accustomed to with plugins. This means re-builds will be automatic when a change occurs as well as giving time to address any developer experience shortcomings which were not foreseen. During this time we will work on overall Bazel documentation as it pertains to the Kibana repository.

> ## Phase III - Plugins

Reviewer 4: ```suggestion
## Phase III - Core & Plugins
```

In this phase, we will be migrating each of the 135 plugins over to being built and unit tested using Bazel. During this time, the legacy systems will stay in place and run in parallel with Bazel. Once all plugins have been migrated, we can decommission the legacy systems.

> The `BUILD.bazel` files will look similar to that of packages, there will be a target for `web`, `server`, and `jest`. Just like packages, as the Jest unit tests are migrated, they will need to be removed from the root `jest.config.js` file as described in the Unit Testing section.

Reviewer 5: Does it mean we are going to have different configuration files for every Environnement? `tsconfig.server.js`, `tsconfig.web.js`, `tsconfig.jest.js`, etc.

Reviewer 3: A lot

Reviewer 3: It only means we will have rules to build web and server separately as we currently have plugins who use different compilers for each target. The configuration part is up to us. Regarding `jest` target it will rely on the previous builds for web and server but it just means we will have a target to run tests all at once for a given package \cc @tylersmalley

Author: Ideally, we only have a single `tsconfig.js` file so we can use its output as the source of multiple targets. However, if we find we need or want separate configurations for different parts we do have that flexibility.

> Plugins are built in a sandbox, so they will no longer be able to use relative imports from one another. For Typescript, relative imports will be replaced with a path reference to the `bazel/bin`.

Reviewer 4: Can we add an example of what this would look like in code? I think how these imports work is an important aspect of how well this tool will fit into our repo. It may also impact how we consider implementing https://github.com/elastic/kibana/issues/71566 and enforce encapsulation between plugins & domains.

Reviewer 6: I agree that we should elaborate on this change and discuss it with a wider audience. A dedicated e-mail to kibana-contributors@ would be great.

Author: Started https://github.com/elastic/kibana/issues/94034 to attempt to get consensus on this: https://github.com/elastic/kibana/issues/94034. Will pass it around to a few folks before blasting kibana-contributors.

Reviewer 6: Thanks, @tylersmalley.

> Static imports across plugins are a concern that would affect the developer experience due to cascading re-builds. For example, if every plugin has static imports from `src/core`, any changes to `src/core` would cause all those plugins to re-build. There are a few options to address this; the first would be to minimize or eliminate these imports. Most plugins are importing types, so we can also ensure that only type-level changes actually trigger a re-build. Additionally, these types of dependencies could be further broken out into another package to reduce the times further this is necessary. 

Reviewer 5: Is Bazel able to understand whether the code or type has been changed? Or we need to annotate import/export explicitly with `import type { ... } from 'src/core/server'`

Reviewer 3: For the best of our knowledge today it looks like the typescript rules for Bazel only triggers cascading rebuilds when the types changed. On the tests I did with @tylersmalley we were not using `import type`.

Reviewer 4: I think this is a bit more clear, if I'm understanding correctly.

```suggestion
Static imports across plugins are a concern that would affect the developer experience due to cascading re-builds. For example, if every plugin has static imports from `src/core`, any changes to `src/core` would cause all those plugins to re-build. There are a few options to address this; the first would be to minimize or eliminate these imports. Most plugins are importing types, so we can also ensure that only type-level changes actually trigger a re-build. Additionally, these types of dependencies could be further broken down into smaller packages to reduce the times further this is necessary. 
```

        "./bazel/out/k8-fastbuild/bin/path/to",
        "./bazel/out/x64_windows-fastbuild/bin/path/to",
        "./bazel/out/darwin-dbg/bin/path/to",
        "./bazel/out/k8-dbg/bin/path/to",
        "./bazel/out/x64_windows-dbg/bin/path/to",
    ]
}
```

## Phase IV -  Build Packaging

In this phase, we will be replacing our current build tooling located at `src/dev/build` to use Bazel. A single target of `release` will provide all assets needed by the release manager:

* Windows (zip)
* Linux 64-bit (tar)
* Linux aarch64 (tar)
* RPM 64-bit
* RPM aarch64
* Deb 64-bit
* Deb aarch64
* Darwin 64-bit (tar)
* CentOS 64-bit Docker Image & Context (tar)
* CentOS aarch64 Docker Image & Context (tar)
* UBI Docker Image & Context (tar)
* Ironbank Docker Context (tar)

There are a few rules already available provided by Bazel that should be used. In some cases, like tar, they have been re-implemented to ensure the output is hermetic. `rules_pgk` has `pkg_tar`, `pkg_deb`, `pkg_rpm`, and `pkg_zip` to assist with this. `rules_docker` provides the ability to build containers without depending on Docker to be installed and providing the ability to build for other platforms.

> While this phase can being with phase 1, it can not be completed until all packages and plugins have been migrated.

Reviewer: ```suggestion
While this phase can begin with phase 1, it can not be completed until all packages and plugins have been migrated.
```

Building TypeScript reference definitions and using `@babel/register` will be negated by using the TypeScript compiler directly instead of using Babel. Currently, we use Babel for code generation and `tsc` for type check and type declaration output. Additionally, the TypeScript implementation in [rules_nodejs](https://bazelbuild.github.io/rules_nodejs/TypeScript.html) for Bazel handles incremental builds, resulting in faster re-builds.

In addition to the benefits of building code, there are also benefits regarding running unit tests. Developers currently need to understand what unit tests to run to validate changes or rely on waiting for CI, which has a long feedback loop. Since Bazel knows the dependency tree, it will only run unit tests for a package or plugin modified or dependent on those modifications. This optimization helps developers and will significantly reduce the amount of work CI needs to do. On CI, unit tests take 35 minutes to complete, where the average single Jest project takes just twenty seconds.

> 

Author: I see no problem with this, and I plan to work towards moving functional tests to have a plugin owner where they will reside. There are a lot of benefits to this, including clear ownership for functional tests to assist with tooling and how we report failures. In addition, it would help get us on a path to reducing the number of functional tests we run _by default_ on a pull request. Bazel provides the ability to [query](https://docs.bazel.build/versions/master/query-how-to.html) the action graph to understand what packages are affected by a change. We can use this, and additional configuration on the plugins to determine which functional tests should be run by default while allowing the PR to request additional tests to be run.

Bazel maintains a persistent worker which `ts_project` takes advantage of by keeping the AST in memory and providing incremental updates. This should improve the time it takes for changes to be represented. 

A Bazel [macro](https://docs.bazel.build/versions/master/skylark/macros.html) will be created to centralize the usage of `ts_project`. The macro will, at minimum, accept a TypeScript configuration file, supply the base `tsconfig.js` file as a source and ensure incremental builds are enabled.

## Webpack

A Bazel [macro](https://docs.bazel.build/versions/master/skylark/macros.html) will be created to centralize the usage of Webpack. The macro will, at minimum, accept a configuration file and supply a base `webpack.config.js` file. Currently, all plugins share the same Webpack configuration. Allowing a plugin to provide additional configuration will allow plugins the ability to add loaders without affecting the performance of others.

> While running Kibana from source in development, the proxy server will ensure that client-side code for plugins is compiled and available. This is handled by the [optimizerMixin](https://github.com/elastic/kibana/blob/master/src/optimize/optimize_mixin.ts), where routes for the client-side assets are created. When code is updated, the routes will pause requests while @kbn/optimizer re-bundles the plugins. With Bazel, we will utilize [iBazel](https://github.com/bazelbuild/bazel-watche) to watch for file changes and re-build the plugin targets when necessary. The watcher will emit [events](https://github.com/bazelbuild/bazel-watcher#remote-events) that we will use to block requests and provide feedback to the logs.

Reviewer 7: ```suggestion
While running Kibana from source in development, the proxy server will ensure that client-side code for plugins is compiled and available. This is currently handled by the [basePathProxy](https://github.com/elastic/kibana/blob/master/src/core/server/http/base_path_proxy_server.ts), where server restarts and optimizer builds are observed and cause the proxy to pause requests. With Bazel, we will utilize [iBazel](https://github.com/bazelbuild/bazel-watche) to watch for file changes and re-build the plugin targets when necessary. The watcher will emit [events](https://github.com/bazelbuild/bazel-watcher#remote-events) that we will use to block requests and provide feedback to the logs.
```

We are aware there are quite a few alternatives to Webpack, but our plan is to continue using it during the migration. Once all packages have been migrated to Bazel, it will be much easier to test alternatives through changing the targets of a single plugins `BUILD.bazel` file.

### Unit Testing

A Bazel macro will be created to centralize the usage of Jest unit testing. The macro will, at minimum, accept a Jest configuration file, add the [Jest preset](https://github.com/elastic/kibana/blob/master/packages/kbn-test/jest-preset.js) and its dependencies as sources, then use the Jest CLI to execute tests. 

Developers currently use `yarn test:jest` to efficiently run tests in a given directory without remembering the command or path. This command will continue to work as it does today, but will begin running tests through Bazel for packages or plugins which have been migrated.

> CI will have an additional job to run `bazel test //…:jest`. This will run unit tests for any package or plugin modified or dependent on modifications since the last successful CI run on that branch.

Reviewer 8: Can you explain "dependent on modifications"

Core (as well as other plugins) maintains many mocks that are used by other plugin's unit tests. Would a change in a mock result in unit tests being run for all the tests that import that mock?

Author: As it currently stands, yes. The mocks are inputs into the test actions, and if those change the downstream tests will be re-run.

Author: There are ways to combat this by isolating the mocks with a separate [filegroup](https://docs.bazel.build/versions/master/be/general.html#filegroup) which would be consumed by the downstream projects.

### Compiling server-side code

While in development, we rely on [@babel/register](https://babel.dev/docs/en/babel-register) to transpile server-side code during runtime. The use of `@babel/register` results in the compile-time cost being paid when the code is run, mostly during startup or when initiating a unit test. When comparing development startup to production, where the code is pre-compiled, we see startup time taking about twice as long even when the Babel cache exists.

---

> These steps cost developers move than ten minutes of their time when updating or changing branches. These times will continue to worsen as the project continues to grow. Instead of making small incremental changes as we have done in the past, like improving caching in a single area, we would like to leverage Bazel, where there are already solutions to many of these problems.

Reviewer 3: typo `move` -> `more`

Author: Resolved.
